File: frontend/src/lib/loose-version.ts
/**
 * Munki-compatible loose version comparison (Python distutils.version.LooseVersion).
 * @see https://managingosx.wordpress.com/2020/10/06/this-one-goes-to-11-macos-version-comparisons-and-munki/
 */

const COMPONENT_RE = /(\d+|[a-z]+|\.)/gi

export type LooseVersionPart = number | string

export function parseLooseVersion(
  vstring: string | null | undefined,
): LooseVersionPart[] {
  if (!vstring) return []
  const components: LooseVersionPart[] = []
  for (const match of vstring.matchAll(COMPONENT_RE)) {
    const part = match[0]
    if (!part || part === '.') continue
    const asNumber = Number(part)
    components.push(Number.isNaN(asNumber) ? part : asNumber)
  }
  return components
}

function compareLooseParts(
  left: LooseVersionPart[],
  right: LooseVersionPart[],
): number {
  const max = Math.max(left.length, right.length)
  for (let i = 0; i < max; i += 1) {
    const l = left[i]
    const r = right[i]
    if (l === undefined) return r === undefined ? 0 : -1
    if (r === undefined) return 1
    if (l === r) continue
    if (typeof l === 'number' && typeof r === 'number') {
      return l < r ? -1 : 1
    }
    const ls = String(l)
    const rs = String(r)
    if (ls < rs) return -1
    if (ls > rs) return 1
  }
  return 0
}

/** Return -1, 0, or 1 for two version strings. */
export function compareLooseVersions(
  left: string | null | undefined,
  right: string | null | undefined,
): number {
  return compareLooseParts(parseLooseVersion(left), parseLooseVersion(right))
}

export function sortByLooseVersion<T>(
  items: T[],
  getVersion: (item: T) => string | null | undefined,
  order: 'asc' | 'desc' = 'desc',
): T[] {
  const sorted = [...items].sort((a, b) =>
    compareLooseVersions(getVersion(a), getVersion(b)),
  )
  return order === 'desc' ? sorted.reverse() : sorted
}

/** TanStack Table sortingFn for version columns. */
export function looseVersionSortingFn<T>(
  rowA: { getValue: (columnId: string) => unknown },
  rowB: { getValue: (columnId: string) => unknown },
  columnId: string,
): number {
  return compareLooseVersions(
    rowA.getValue(columnId) as string | null | undefined,
    rowB.getValue(columnId) as string | null | undefined,
  )
}

File: backend/automunki/services/loose_version.py
"""Munki-compatible loose version comparison (distutils.version.LooseVersion)."""

from __future__ import annotations

import re
from functools import total_ordering
from typing import Any

_COMPONENT_RE = re.compile(r"(\d+|[a-z]+|\.)", re.VERBOSE)


def parse_loose_version(vstring: str | None) -> list[int | str]:
    """Split a version string into comparable components."""
    if not vstring:
        return []
    components: list[int | str] = []
    for part in _COMPONENT_RE.split(vstring):
        if not part or part == ".":
            continue
        try:
            components.append(int(part))
        except ValueError:
            components.append(part)
    return components


def compare_loose_versions(left: str | None, right: str | None) -> int:
    """Return -1, 0, or 1 (like ``cmp``) for two version strings."""
    l_parts = parse_loose_version(left)
    r_parts = parse_loose_version(right)
    if l_parts == r_parts:
        return 0
    if l_parts < r_parts:
        return -1
    return 1


def loose_version_key(vstring: str | None) -> tuple[Any, ...]:
    """Sort key for ``sorted(..., key=loose_version_key)``."""
    return tuple(parse_loose_version(vstring))


@total_ordering
class LooseVersion:
    """Thin wrapper matching Munki's loose version ordering."""

    def __init__(self, vstring: str | None) -> None:
        self.vstring = vstring or ""
        self.version = parse_loose_version(vstring)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, LooseVersion):
            return NotImplemented
        return self.version == other.version

    def __lt__(self, other: LooseVersion) -> bool:
        return self.version < other.version
